README.md
## Setup
1.  Go to the ESO documents directory at ```C:\Users\YOURUSER\Documents\Elder Scrolls Online\VERSION\``` (replace VERSION with the client you're using)
2.  If a folder called ``Addons`` doesn't exist, create it.
3.  Clone this repository inside the ```Addons``` folder as ```Esohead```
4.  Load the game, log into character selection and click the ```Add-Ons``` option. You should see ```Esohead Looter``` on your list, please ensure ```Load Out of Date Addons``` is selected if the addon is out of date.
5.  After logging into the game, you should see ```Esohead addon initialized. Debugging is disabled.```

## SavedVariables
After logging into the game and gathering some data, and logging out completely or typing ```/reloadui```, a SavedVariables file may be examined at ```C:\Users\YOURUSER\Documents\Elder Scrolls Online\VERSION\SavedVariables\Esohead.lua```

## Slash Commands
```/esohead reset```
Completely resets all gathered data.

```/esohead reset DATATYPE```
Resets a specific type of data

```/esohead debug on|off```
Toggles verbose addon debugging.

```/esohead datalog```
Displays the total number of each data type the addon has gathered.

```/rl``` & ```/reload```
Aliases of the reloadui command

## Core Functions
**EH.Log** _(**string** type, **object** nodes, ...)_<br />
Logs any type of data to the Esohead Saved Variables file. Takes an ordered table of strings that define the node
and/or sub-nodes that data will be logged to.

For example
```lua
EH.Log("npc", { "Glenumbra", "Covenant Archer" }, 0.67, 0.47)
```
Would create an entry in saved variables like this
```lua
Esohead_SavedVariables =
{
    ["Default"] =
    {
        ["@mdurrant"] =
        {
            ["$AccountWide"] =
            {
                ["npc"] =
                {
                    ["Glenumbra"] =
                    {
                        ["Covenant Archer"] =
                        {
                            [1] =
                            {
                                [1] = 0.67,
                                [2] = 0.47,
                            },
                        },
                    },
                },
            },
        },
    },
},
```
<br />
**EH.LogCheck** _(**string** type, **object** nodes, x, y)_<br />
Checks an ordered table of nodes for an x and y position, returns false if there is an entry that is close to those coordinates.

**EH.NumberFormat** _(**int** number)_<br />
Returns a comma-formatted number string for display to the user.

**EH.Debug** <br />
Displays a message in the chat window containing the output of the data passed to it. Accepts any data type or function.


## API Helpers
**EH.GetUnitPosition** _(**string** unitTag)_<br />
_returns **float** x, **float** y, **float** heading, **string** subZone, **string** world_<br /><br />
Combines 3 API commands and returns all the crucial data to pinpointing a location.

## Disclaimer

This Add-on is not created by, affiliated with or sponsored by ZeniMax Media Inc. or its affiliates. The Elder Scrolls® and related logos are registered trademarks or trademarks of ZeniMax Media Inc. in the United States and/or other countries. All rights reserved.


CHANGELOG.md
##0.1.14

Updates
- Updated TOC; Checked change notes no routines were effected

##0.1.13

(changes made by Seerah)

Bugfixes
- You cannot get position information for the reticleover unit any longer,
	can only collect information of npcs that you interact with, and use the player position.
	Moved the NPC logging code to the EH.OnChatterBegin function.
	
Changes
- Removed residual provisioning tracking code

##0.1.12

Minor Update
- Sharlikran PR: Updated TOC

##0.1.11

Bugfixes
- Sharlikran PR: Fixed checking for duplicate data entries

##0.1.10

Bugfixes
- Data logged with double quotes in the name are now escaped properly, this addresses many lua errors users were experiencing.
- Sharlikran PR#4: Various fixes with harvest loot accuracy
- Fixed vendor tracking
- Removed debug message for harvesting Nodes that do not exist in EsoheadConstants.lua, it is no longer needed.

Changes
- Removed logging of provisioning data

Features
- User can reset specified data type (/esohead reset npc)

Credits:
- Code snippets borrowed from HarvestMap by Shinni

##0.1.9

Bugfixes
- Sharlikran PR#4: Added "Heavy Sack" and changed "Kresh Weed" and "Silver Weed" to "Kreshweed" and "Silverweed"

Features
- Added localization for German and French.  While German only needs a few translations, the French localization needs many updates.  Any help would be appreciated.
- Added debug message for harvesting Nodes that do not exist in EsoheadConstants.lua.  It should only fire when you are harvesting.

Credits:
- German Localization: KinqxsYrox, and Rathputin
- French Localization: Deadlock, jillorval, Kalmeth, and wookiefrag

##0.1.8

Bugfixes
- Sharlikran PR: Changed Node Name Case and Added "Kresh Weed", "Pure Water", "Silver Weed"

##0.1.7

Bugfixes
- Map callbacks were being fired way more than necessary from a fix in 0.1.6, causing FPS drops whenever a reticle hovers over a loggable object. This has been addressed to fire only when it should.

##0.1.6

Bugfixes
- Fixed a bug where a coordinates would be recorded on the wrong map if the player navigates away from the current map. Bug #23 by Shinni, ref: http://www.esoui.com/portal.php?id=1&a=viewbug&bugid=23
- Increased API version to 100003 to resolve out of date issues.

##0.1.5

Bugfixes
- Harvest/Provisioning data is now being collected when the player has Auto Loot enabled.

Features
- ``EH.lastTarget`` has been added to provide the name of the last NPC/Object the player had their reticle over _before_ any interaction commands are fired. This is useful, for example, in having the name of the harvest node available immediately after the harvest is complete (when EVENT_LOOT_RECEIVED is fired on auto-loot), even if the player's reticle wanders around and targets other things while they're harvesting.
- A new function ``EH.IsValidNode`` has been added to support an additional table of game data generated in our EsoheadConstants.lua file. This function can be used to pass the name of an interactable object and determine if it's a valid harvesting node (includes provisioning nodes).

##0.1.4

Bugfixes
- Added an additional check to ensure a material is a provisioning-type before logging it
- Unexpectedly empty fields should no longer be stored

Features
- Improved the reliability of the ``EH.currentTarget`` variable
- Narrowed the distance a player needs to be from a previously logged data point in order to log a nearby point from 1% map size to 0.5%

##0.1.3

Bugfixes
- Fixed a bug where looting a tradeskill material from an enemy counted as a harvest node

Features
- The addon is now called "Esohead" rather than "Esohead Looter"
- Saved variables for each data type are now independently versioned from eachother. This allows us to change the format data is collected and force the client to clear out the old version's obsolete data.
- Provisioning materials are now logged differently from materials of other tradeskills.

##0.1.2

Bugfixes
- Harvest data is no longer collected when a player opens a container in their inventory with tradeskill materials (such as from a hireling).

##0.1.1

Features
- Complete refactor of the addon, it is no longer a ZO_CallbackObject
- ``EH.ItemLinkParse`` has been added to parse in-game item links for ingesting key parts of the string.

Esohead.lua
-----------------------------------------
--                                     --
--            E s o h e a d            --
--                                     --
--    E-mail: feedback@esohead.com     --
--                                     --
-----------------------------------------

EH = {}

-----------------------------------------
--           Core Functions            --
-----------------------------------------

function EH.Initialize()
    EH.savedVars = {}
    EH.debugDefault = 0
    EH.dataDefault = {
        data = {}
    }

    EH.name = ""
    EH.time = 0
    EH.isHarvesting = false
    EH.action = ""

    EH.currentConversation = {
        npcName = "",
        npcLevel = 0,
        x = 0,
        y = 0,
        subzone = ""
    }
    EH.minDefault = 0.000025 -- 0.005^2
    EH.minReticleover = 0.000049 -- 0.007^2
end

function EH.InitSavedVariables()
    EH.savedVars = {
        ["internal"]     = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 1, "internal", { debug = EH.debugDefault, language = "" }),
        ["skyshard"]     = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "skyshard", EH.dataDefault),
        ["book"]         = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "book", EH.dataDefault),
        ["harvest"]      = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 4, "harvest", EH.dataDefault),
        --[[ ["provisioning"] = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 5, "provisioning", EH.dataDefault), ]]--
        ["chest"]        = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "chest", EH.dataDefault),
        ["fish"]         = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "fish", EH.dataDefault),
        ["npc"]          = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "npc", EH.dataDefault),
        ["vendor"]       = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "vendor", EH.dataDefault),
        ["quest"]        = ZO_SavedVars:NewAccountWide("Esohead_SavedVariables", 2, "quest", EH.dataDefault),
    }

    if EH.savedVars["internal"].debug == 1 then
        EH.Debug("Esohead addon initialized. Debugging is enabled.")
    else
        EH.Debug("Esohead addon initialized. Debugging is disabled.")
    end
end

-- Logs saved variables
function EH.Log(type, nodes, ...)
    local data = {}
    local dataStr = ""
    local sv

    if EH.savedVars[type] == nil or EH.savedVars[type].data == nil then
        EH.Debug("Attempted to log unknown type: " .. type)
        return
    else
        sv = EH.savedVars[type].data
    end

    for i = 1, #nodes do
        local node = nodes[i];
        if string.find(node, '\"') then
            node = string.gsub(node, '\"', '\'')
        end

        if sv[node] == nil then
            sv[node] = {}
        end
        sv = sv[node]
    end

    for i = 1, select("#", ...) do
        local value = select(i, ...)
        data[i] = value
        dataStr = dataStr .. "[" .. tostring(value) .. "] "
    end

    if EH.savedVars["internal"].debug == 1 then
        EH.Debug("Logged [" .. type .. "] data: " .. dataStr)
    end

    if #sv == 0 then
        sv[1] = data
    else
        sv[#sv+1] = data
    end
end

-- Checks if we already have an entry for the object/npc within a certain x/y distance
function EH.LogCheck(type, nodes, x, y, scale, name)
    local log = true
    local sv

    if x <= 0 or y <= 0 then
        return false
    end

    if EH.savedVars[type] == nil or EH.savedVars[type].data == nil then
        return true
    else
        sv = EH.savedVars[type].data
    end

    local distance
    if scale == nil then
        distance = EH.minDefault
    else
        distance = scale
    end

    for i = 1, #nodes do
        local node = nodes[i];
        if string.find(node, '\"') then
            node = string.gsub(node, '\"', '\'')
        end

        if sv[node] == nil then
            sv[node] = {}
        end
        sv = sv[node]
    end

    for i = 1, #sv do
        local item = sv[i]

        dx = item[1] - x
        dy = item[2] - y
        -- (x - center_x)2 + (y - center_y)2 = r2, where center is the player
        dist = math.pow(dx, 2) + math.pow(dy, 2)
        -- both ensure that the entire table isn't parsed
        if dist < distance then -- near player location
            if name == nil then -- npc, quest, vendor all but harvesting
                return false
            else -- harvesting only
                if item[4] == name then
                    return false
                elseif item[4] ~= name then
                    return true
                end
            end
        end
    end

    return log
end

-- formats a number with commas on thousands
function EH.NumberFormat(num)
    local formatted = num
    local k

    while true do
        formatted, k = string.gsub(formatted, "^(-?%d+)(%d%d%d)", '%1,%2')
        if k == 0 then
            break
        end
    end

    return formatted
end

-- Listens for anything that is not event driven by the API but needs to be tracked
function EH.OnUpdate(time)
    if IsGameCameraUIModeActive() or IsUnitInCombat("player") then
        return
    end

    local type = GetInteractionType()
    local active = IsPlayerInteractingWithObject()
    local x, y, a, subzone, world = EH.GetUnitPosition("player")
    local targetType
    local action, name, interactionBlocked, additionalInfo, context = GetGameCameraInteractableActionInfo()

    local isHarvesting = ( active and (type == INTERACTION_HARVEST) )
    if not isHarvesting then
        if name then
            EH.name = name -- EH.name is the global current node
        end

        if EH.isHarvesting and time - EH.time > 1 then
            EH.isHarvesting = false
        end

        if action ~= EH.action then
            EH.action = action -- EH.action is the global current action

            -- Check Reticle and Non Harvest Actions
            -- Skyshard
            if type == INTERACTION_NONE and EH.action == GetString(SI_GAMECAMERAACTIONTYPE5) then
                targetType = "skyshard"

                if EH.name == "Skyshard" then
                    if EH.LogCheck(targetType, {subzone}, x, y, EH.minReticleover, nil) then
                        EH.Log(targetType, {subzone}, x, y)
                    end
                end

            -- Chest
            elseif type == INTERACTION_NONE and EH.action == GetString(SI_GAMECAMERAACTIONTYPE12) then
                targetType = "chest"

                if EH.LogCheck(targetType, {subzone}, x, y, EH.minReticleover, nil) then
                    EH.Log(targetType, {subzone}, x, y)
                end

            -- Fishing Nodes
            elseif EH.action == GetString(SI_GAMECAMERAACTIONTYPE16) then
                targetType = "fish"

                if EH.LogCheck(targetType, {subzone}, x, y, EH.minReticleover, nil) then
                    EH.Log(targetType, {subzone}, x, y)
                end

            end
        end
    else
        EH.isHarvesting = true
        EH.time = time

    end
end

-----------------------------------------
--         Coordinate System           --
-----------------------------------------

function EH.UpdateCoordinates()
    local mouseOverControl = WINDOW_MANAGER:GetMouseOverControl()

    if (mouseOverControl == ZO_WorldMapContainer or mouseOverControl:GetParent() == ZO_WorldMapContainer) then
        local currentOffsetX = ZO_WorldMapContainer:GetLeft()
        local currentOffsetY = ZO_WorldMapContainer:GetTop()
        local parentOffsetX = ZO_WorldMap:GetLeft()
        local parentOffsetY = ZO_WorldMap:GetTop()
        local mouseX, mouseY = GetUIMousePosition()
        local mapWidth, mapHeight = ZO_WorldMapContainer:GetDimensions()
        local parentWidth, parentHeight = ZO_WorldMap:GetDimensions()

        local normalizedX = math.floor((((mouseX - currentOffsetX) / mapWidth) * 100) + 0.5)
        local normalizedY = math.floor((((mouseY - currentOffsetY) / mapHeight) * 100) + 0.5)

        EsoheadCoordinates:SetAlpha(0.8)
        EsoheadCoordinates:SetDrawLayer(ZO_WorldMap:GetDrawLayer() + 1)
        EsoheadCoordinates:SetAnchor(TOPLEFT, nil, TOPLEFT, parentOffsetX + 0, parentOffsetY + parentHeight)
        EsoheadCoordinatesValue:SetText("Coordinates: " .. normalizedX .. ", " .. normalizedY)
    else
        EsoheadCoordinates:SetAlpha(0)
    end
end

-----------------------------------------
--            API Helpers              --
-----------------------------------------

function EH.GetUnitPosition(tag)
    local setMap = SetMapToPlayerLocation() -- Fix for bug #23
    if setMap == 2 then
        CALLBACK_MANAGER:FireCallbacks("OnWorldMapChanged") -- Fix for bug #23
    end

    local x, y, a = GetMapPlayerPosition(tag)
    local subzone = GetMapName()
    local world = GetUnitZone(tag)

    return x, y, a, subzone, world
end

function EH.GetUnitName(tag)
    return GetUnitName(tag)
end

function EH.GetUnitLevel(tag)
    return GetUnitLevel(tag)
end

function EH.GetLootEntry(index)
    return GetLootItemInfo(index)
end

-----------------------------------------
--           Debug Logger              --
-----------------------------------------

local function EmitMessage(text)
    if(CHAT_SYSTEM)
    then
        if(text == "")
        then
            text = "[Empty String]"
        end

        CHAT_SYSTEM:AddMessage(text)
    end
end

local function EmitTable(t, indent, tableHistory)
    indent          = indent or "."
    tableHistory    = tableHistory or {}

    for k, v in pairs(t)
    do
        local vType = type(v)

        EmitMessage(indent.."("..vType.."): "..tostring(k).." = "..tostring(v))

        if(vType == "table")
        then
            if(tableHistory[v])
            then
                EmitMessage(indent.."Avoiding cycle on table...")
            else
                tableHistory[v] = true
                EmitTable(v, indent.."  ", tableHistory)
            end
        end
    end
end

function EH.Debug(...)
    for i = 1, select("#", ...) do
        local value = select(i, ...)
        if(type(value) == "table")
        then
            EmitTable(value)
        else
            EmitMessage(tostring (value))
        end
    end
end

-----------------------------------------
--           Loot Tracking             --
-----------------------------------------

function EH.ItemLinkParse(link)

    local Field1, Field2, Field3, Field4, Field5 = ZO_LinkHandler_ParseLink( link )

    -- name = Field1
    -- unused = Field2
    -- type = Field3
    -- id = Field4
    -- quality = Field5

    return {
        type = Field3,
        id = tonumber(Field4),
        quality = tonumber(Field5),
        name = zo_strformat(SI_TOOLTIP_ITEM_NAME, Field1)
    }
end

function EH.OnLootReceived(eventCode, receivedBy, objectName, stackCount, soundCategory, lootType, lootedBySelf)
    if not IsGameCameraUIModeActive() then
        targetName = EH.name
        
        -- There is no provisioning now so if the player isn't harvesting
        -- then you don't need to do anything.
        if not EH.isHarvesting then 
            return
        end

        -- if not EH.IsValidNode(targetName) then
        --     return
        -- end

        local link = EH.ItemLinkParse(objectName)
        local material = EH.GetTradeskillByMaterial(link.id)
        local x, y, a, subzone, world = EH.GetUnitPosition("player")

        --[[
        if not EH.isHarvesting and material >= 1 then
            material = 5
        elseif EH.isHarvesting and material == 5 then
            material = 0
        end
        ]]--

        if material == 0 then
            return
        end

        --[[ If provisioning is used in the future, change EH.LogCheck to return the item so it can be inserted into the table
        if material == 5 then
            data = EH.LogCheck("provisioning", {subzone, material, link.id}, x, y, nil)
            if not data then -- when there is no node at the given location, save a new entry
                EH.Log("provisioning", {subzone, material, link.id}, x, y, stackCount, targetName)
        else
        ]]--
            if EH.LogCheck("harvest", {subzone, material}, x, y, nil, targetName) then
                EH.Log("harvest", {subzone, material}, x, y, stackCount, targetName, link.id)
            end
   --[[ end ]]--
    end
end

-----------------------------------------
--         Lore Book Tracking          --
-----------------------------------------

function EH.OnShowBook(eventCode, title, body, medium, showTitle)
    local x, y, a, subzone, world = EH.GetUnitPosition("player")

    local targetType = "book"

    if EH.LogCheck(targetType, {subzone, title}, x, y, nil, nil) then
        EH.Log(targetType, {subzone, title}, x, y)
    end
end

-----------------------------------------
--          Vendor Tracking            --
-----------------------------------------

function EH.VendorOpened()
    local x, y, a, subzone, world = EH.GetUnitPosition("player")

    targetType = "vendor"

    local storeItems = {}

    if EH.LogCheck(targetType, {subzone, EH.name}, x, y, 0.1, nil) then
        for entryIndex = 1, GetNumStoreItems() do
            local icon, name, stack, price, sellPrice, meetsRequirementsToBuy, meetsRequirementsToEquip, quality, questNameColor, currencyType1, currencyId1, currencyQuantity1, currencyIcon1,
            currencyName1, currencyType2, currencyId2, currencyQuantity2, currencyIcon2, currencyName2 = GetStoreEntryInfo(entryIndex)

            if(stack > 0) then
            local itemData =
            {
                name,
                stack,
                price,
                quality,
                questNameColor,
                currencyType1,
                currencyQuantity1,
                currencyType2,
                currencyQuantity2,
                { GetStoreEntryTypeInfo(entryIndex) },
                GetStoreEntryStatValue(entryIndex),
                }

                storeItems[#storeItems + 1] = itemData
            end
        end

        EH.Log(targetType, {subzone, EH.name}, x, y, storeItems)
    end
end

-----------------------------------------
--           Quest Tracking            --
-----------------------------------------

function EH.OnQuestAdded(_, questIndex)
    -- This routine might need a check to make sure the player 
    -- obtained the quest by interacting with the NPC and that
    -- the quest was not shared by another player
    local questName = GetJournalQuestInfo(questIndex)
    local questLevel = GetJournalQuestLevel(questIndex)

    local targetType = "quest"

    if EH.currentConversation.npcName == "" or EH.currentConversation.npcName == nil then
        return
    end
    -- quests are not tracked when your reticle is over the quest giver
    -- however the NPC might move around, therefore range checking in needed
    if EH.LogCheck(targetType, {EH.currentConversation.subzone, questName}, EH.currentConversation.x, EH.currentConversation.y, EH.minReticleover, nil) then
        EH.Log(
            targetType,
            {
                EH.currentConversation.subzone,
                questName
            },
            EH.currentConversation.x,
            EH.currentConversation.y,
            questLevel,
            EH.currentConversation.npcName,
            EH.currentConversation.npcLevel
        )
    end
end

-----------------------------------------
--        Conversation Tracking        --
-----------------------------------------

function EH.OnChatterBegin()
    local x, y, a, subzone, world = EH.GetUnitPosition("player")
    local npcLevel = EH.GetUnitLevel("interact")

    EH.currentConversation.npcName = EH.name
    EH.currentConversation.npcLevel = npcLevel
    EH.currentConversation.x = x
    EH.currentConversation.y = y
    EH.currentConversation.subzone = subzone
	
	-- You cannot get position information for the reticleover unit any longer.
	-- Can only collect information of npcs that you interact with, and use the player position.
	-- Moved this logging code to this function.
	local name = EH.currentConversation.npcName
	if EH.LogCheck("npc", {subzone, name}, x, y, EH.minReticleover, nil) then
		EH.Log("npc", {subzone, name}, x, y, npcLevel)
	end

end

-----------------------------------------
--        Better NPC Tracking          --
-----------------------------------------

-- Fired when the reticle hovers a new target

-- You cannot get position information for the reticleover unit any longer.
-- Can only collect information of npcs that you interact with, and use the player position.
-- Moved the logging code to the EH.OnChatterBegin function.

--[[function EH.OnTargetChange(eventCode)
    local tag = "reticleover"
    local type = GetUnitType(tag)

    -- ensure the unit that the reticle is hovering is a non-playing character
    if type == 2 then
        local name = EH.GetUnitName(tag)
        local x, y, a, subzone, world = EH.GetUnitPosition(tag)

        if name == nil or name == "" or x <= 0 or y <= 0 then
            return
        end

        local level = EH.GetUnitLevel(tag)

		if EH.LogCheck("npc", {subzone, name }, x, y, EH.minReticleover, nil) then
			EH.Log("npc", {subzone, name}, x, y, level)
		end
	end
end]]

-----------------------------------------
--           Slash Command             --
-----------------------------------------

EH.validCategories = {
    "chest",
    "fish",
    --[[ "provisioning", ]]--
    "book",
    "vendor",
    "quest",
    "harvest",
    "npc",
    "skyshard",
}

function EH.IsValidCategory(name)
    for k, v in pairs(EH.validCategories) do
        if string.lower(v) == string.lower(name) then
            return true
        end
    end

    return false
end

SLASH_COMMANDS["/esohead"] = function (cmd)
    local commands = {}
    local index = 1
    for i in string.gmatch(cmd, "%S+") do
        if (i ~= nil and i ~= "") then
            commands[index] = i
            index = index + 1
        end
    end

    if #commands == 0 then
        return EH.Debug("Please enter a valid Esohead command")
    end

    if #commands == 2 and commands[1] == "debug" then
        if commands[2] == "on" then
            EH.Debug("Esohead debugger toggled on")
            EH.savedVars["internal"].debug = 1
        elseif commands[2] == "off" then
            EH.Debug("Esohead debugger toggled off")
            EH.savedVars["internal"].debug = 0
        end

    elseif commands[1] == "reset" then
        if #commands ~= 2 then 
            for type,sv in pairs(EH.savedVars) do
                if type ~= "internal" then
                    EH.savedVars[type].data = {}
                end
            end
            EH.Debug("Esohead saved data has been completely reset")
        else
            if commands[2] ~= "internal" then
                if EH.IsValidCategory(commands[2]) then
                    EH.savedVars[commands[2]].data = {}
                    EH.Debug("Esohead saved data : " .. commands[2] .. " has been reset")
                else
                    EH.Debug("Please enter a valid Esohead category to reset.")
                    EH.Debug("Valid categories are chest, fish, book, vendor,") 
                    EH.Debug("quest, harvest, npc, and skyshard.")
                    return
                end
            end
        end

    elseif commands[1] == "datalog" then
        EH.Debug("---")
        EH.Debug("Complete list of gathered data:")
        EH.Debug("---")

        local counter = {
            ["skyshard"] = 0,
            ["npc"] = 0,
            ["harvest"] = 0,
            --[[ ["provisioning"] = 0, ]]--
            ["chest"] = 0,
            ["fish"] = 0,
            ["book"] = 0,
            ["vendor"] = 0,
            ["quest"] = 0,
        }

        for type,sv in pairs(EH.savedVars) do
            if type ~= "internal" and (type == "skyshard" or type == "chest" or type == "fish") then
                for zone, t1 in pairs(EH.savedVars[type].data) do
                    counter[type] = counter[type] + #EH.savedVars[type].data[zone]
                end
            --[[ elseif type ~= "internal" and type == "provisioning" then
                for zone, t1 in pairs(EH.savedVars[type].data) do
                    for item, t2 in pairs(EH.savedVars[type].data[zone]) do
                        for data, t3 in pairs(EH.savedVars[type].data[zone][item]) do
                            counter[type] = counter[type] + #EH.savedVars[type].data[zone][item][data]
                        end
                    end
                end ]]--
            elseif type ~= "internal" then
                for zone, t1 in pairs(EH.savedVars[type].data) do
                    for data, t2 in pairs(EH.savedVars[type].data[zone]) do
                        counter[type] = counter[type] + #EH.savedVars[type].data[zone][data]
                    end
                end
            end
        end

        EH.Debug("Skyshards: "        .. EH.NumberFormat(counter["skyshard"]))
        EH.Debug("Monster/NPCs: "     .. EH.NumberFormat(counter["npc"]))
        EH.Debug("Lore/Skill Books: " .. EH.NumberFormat(counter["book"]))
        EH.Debug("Harvest: "          .. EH.NumberFormat(counter["harvest"]))
        --[[ EH.Debug("Provisioning: "     .. EH.NumberFormat(counter["provisioning"])) ]]--
        EH.Debug("Treasure Chests: "  .. EH.NumberFormat(counter["chest"]))
        EH.Debug("Fishing Pools: "    .. EH.NumberFormat(counter["fish"]))
        EH.Debug("Quests: "           .. EH.NumberFormat(counter["quest"]))
        EH.Debug("Vendor Lists: "     .. EH.NumberFormat(counter["vendor"]))

        EH.Debug("---")
    end
end

SLASH_COMMANDS["/rl"] = function()
    ReloadUI("ingame")
end

SLASH_COMMANDS["/reload"] = function()
    ReloadUI("ingame")
end

-----------------------------------------
--        Addon Initialization         --
-----------------------------------------

function EH.OnLoad(eventCode, addOnName)
    if addOnName ~= "Esohead" then
        return
    end

    EH.language = (GetCVar("language.2") or "en")
    EH.InitSavedVariables()
    EH.savedVars["internal"]["language"] = EH.language

    --EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_RETICLE_TARGET_CHANGED, EH.OnTargetChange)
    EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_CHATTER_BEGIN, EH.OnChatterBegin)
    EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_SHOW_BOOK, EH.OnShowBook)
    EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_QUEST_ADDED, EH.OnQuestAdded)
    EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_LOOT_RECEIVED, EH.OnLootReceived)
    -- {{added}}
    EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_OPEN_STORE, EH.VendorOpened)
end

EVENT_MANAGER:RegisterForEvent("Esohead", EVENT_ADD_ON_LOADED, function (eventCode, addOnName)
    if addOnName == "Esohead" then
        EH.Initialize()
        EH.OnLoad(eventCode, addOnName)
    end
end)